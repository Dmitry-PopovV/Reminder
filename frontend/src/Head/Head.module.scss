.container {
    background-color: lightblue;
    width: 100%;
}

.logout_button {
    height: 90%;
}
