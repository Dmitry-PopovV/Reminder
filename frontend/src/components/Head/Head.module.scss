.container {
    background-color: lightblue;
}

.center {
    display: flex;
    align-items: center;
}

.logoImg {
    @media (min-width: 600px) {
        width: 48px;
        height: 48px;
    }

    @media (max-width: 600px) {
        width: 24px;
        height: 24px;
    }
}

.logoTitle {
    @media (min-width: 600px) {
        font-size: 25px;
    }

    font-weight: 500;
}

.login {
    margin-right: 4px;

    @media (max-width: 600px) {
        font-size: 9px;
        display: none;
    }
}
