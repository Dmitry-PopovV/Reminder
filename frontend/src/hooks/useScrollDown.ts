import { useEffect } from 'react';

export function useScrollDown<T>(onChange: T) {
    useEffect(() => {
        window.scrollTo({ top: document.body.scrollHeight, behavior: "instant" });
    }, [onChange])
}
