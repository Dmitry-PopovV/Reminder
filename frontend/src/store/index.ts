import { configureStore } from "@reduxjs/toolkit";
import userReducer from "./slicers/userSlice";
import { useSelector } from "react-redux";

export const store = configureStore({
    reducer: {
        user: userReducer
    }
})
export type RootState = ReturnType<typeof store.getState>
export type AppDispatch = typeof store.dispatch
export const useAppSelector = () => useSelector((state: RootState) => state.user.user);
