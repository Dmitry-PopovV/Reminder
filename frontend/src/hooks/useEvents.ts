import { useEffect } from 'react';
import axios from 'axios';
import { useAppSelector, useAppDispatch } from "../store";
import { Event, setEvents, addEvent, deleteEventById } from "../store/slicers/eventsSlice"

function getInitialLoadedMonths() {
    const currentDate = new Date(Date.now());
    const month = currentDate.getMonth();
    const year = currentDate.getFullYear();
    return [
        `${year - (month - 1 < 0 ? 1 : 0)}-${(month - 1) % 12}`,
        `${year}-${month}`,
        `${year + (month + 1 > 12 ? 1 : 0)}-${(month + 1) % 12}`
    ];
}

export function useEvents() {
    const events = useAppSelector((state) => state.events.events);
    const dispatch = useAppDispatch();
    let loadedMonths: string[] = [];
    if(events) {
        loadedMonths = Object.keys(events);
    }

    useEffect(() => {
        if (!events) {
            axios.post<Event[]>("/api/events", getInitialLoadedMonths())
                .then((res) => {
                    dispatch(setEvents(res.data));
                })
                .catch(() => {
                    dispatch(setEvents([]));
                });
        }
    }, [])

    function newEvent(event: Event) {
        dispatch(addEvent(event));
    }

    function deleteEvent(id: string) {
        dispatch(deleteEventById(id));
    }

    function newMonth(month: string) {
        axios.post<Event[]>("/api/events", [month])
                .then((res) => {
                    dispatch(addEvent(res.data[0]));
                });
    }

    return { events: [{ id: "0", title: "Event", start: "2023-10-28" }], newEvent, deleteEvent, newMonth };
}
