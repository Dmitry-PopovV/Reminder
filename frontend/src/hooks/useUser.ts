import { useEffect } from 'react';
import axios from 'axios';
import { useAppSelector, useAppDispatch } from "../store";
import { User, setUser as setStoreUser } from "../store/slicers/userSlice"

export function useUser() {
  const user = useAppSelector((state) => state.user.user);
  const dispatch = useAppDispatch();

  useEffect(() => {
    if (user === undefined) {
      axios.post("/api/user")
        .then((res) => {
          setUser(res.data)
        })
        .catch(() => {
          setUser(null);
        });
    }
  }, [])

  function setUser(user: User | null) {
    dispatch(setStoreUser(user));
  }

  return { user, setUser };
}
