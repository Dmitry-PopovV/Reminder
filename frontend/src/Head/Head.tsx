import style from "./Head.module.scss";
import Row from 'react-bootstrap/Row';
import Col from 'react-bootstrap/Col';
import Button from 'react-bootstrap/Button';
import axios from 'axios';
import { useUser } from "../hooks/useUser";


export default function Head() {
    const { user, setUser } = useUser();

    function logout() {
        axios.put("/api/auth/logout")
            .then(() => {
                setUser(null);
            })
    }

    return (
        <div className={style.container}>
            <Row className="m-0">
                <Col xs="auto" md="auto">
                    Reminder
                </Col>
                <Col></Col>
                <Col xs="auto" md="auto">
                    <span id="user_name">{user?.email}</span>
                    <Button variant="light" className={style.logout_button} onClick={logout}>logout</Button>
                </Col>
            </Row>
        </div>
    );
}
