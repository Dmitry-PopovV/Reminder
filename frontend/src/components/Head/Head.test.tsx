import { render, screen, fireEvent } from "@testing-library/react";
import { vi } from "vitest";
import Head from "./Head";
import { Provider } from "react-redux";
import axios from "axios";
import { store } from '../../store';
import { useUser } from "../../hooks/useUser";

vi.mock("axios");
vi.mock("../../hooks/useUser");

describe("Head component", () => {
    beforeAll(()=>{
        (axios.put as jest.Mock).mockReturnValue(new Promise((res) => {
            res({
                data: null
            })
        }))
        const setUser = vi.fn();
        (useUser as jest.Mock).mockReturnValue({
            user: {
                email: "test@mail.com",
                fullName: "Test User"
            },
            setUser
        })
    })

    afterEach(()=>{
        vi.clearAllMocks();
    })

    test("snapshot", () => {
        render(
            <Provider store={store}>
                <Head />
            </Provider>
        );
        const head = screen.getByTestId("headMainContainer");
        expect(head).toMatchSnapshot()
    })

    test("logout", () => {
        render(
            <Provider store={store}>
                <Head />
            </Provider>
        );
        const logoutButton = screen.getByTestId("logoutButton");
        fireEvent.click(logoutButton);
        expect(axios.put).toHaveBeenCalledTimes(1);
    })
})
