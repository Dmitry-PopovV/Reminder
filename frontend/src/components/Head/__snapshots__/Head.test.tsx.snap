// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Head component > snapshot 1`] = `
<div
  class="_container_dd69d5 px-5"
  data-testid="headMainContainer"
>
  <div
    class="m-0 justify-content-end row"
  >
    <div
      class="_center_dd69d5 col-md-auto col-auto"
    >
      <img
        class="_logoImg_dd69d5 me-3"
        src="/ReminderLogo512.png"
      />
      <span
        class="_logoTitle_dd69d5"
      >
        Reminder
      </span>
    </div>
    <div
      class="col"
    />
    <div
      class="_center_dd69d5 col-md-auto col-auto"
    >
      <span
        class="_login_dd69d5"
        id="user_name"
      >
        test@mail.com
      </span>
      <button
        class="btn-sm btn btn-secondary"
        data-testid="logoutButton"
        type="button"
      >
        Log out
      </button>
    </div>
  </div>
</div>
`;
